### PR TITLE
[SCFToCalyx] [StaticLogicToCalyx] Pull out utility functions.

### PR DESCRIPTION
**Pull request by cgyurgyik**
Note that some of these may be more appropriate as static helper functions in lowering utils. However, I've made them public for now. Many of the classes are pretty intertwined (component state, lowering state, ...), so I'm trying to tackle some of the smaller pieces to make this transition easier.

### CODE REVIEW

**Comment by cgyurgyik** [comment]
```
FAILED: lib/libCIRCTCalyxTransforms.so 
: && /usr/bin/clang++ -fPIC -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG  -Wl,-z,defs -Wl,-z,nodelete -fuse-ld=lld -Wl,--color-diagnostics   -Wl,--gc-sections -shared -Wl,-soname,libCIRCTCalyxTransforms.so -o lib/libCIRCTCalyxTransforms.so tools/circt/lib/Dialect/Calyx/Transforms/CMakeFiles/obj.CIRCTCalyxTransforms.dir/CompileControl.cpp.o tools/circt/lib/Dialect/Calyx/Transforms/CMakeFiles/obj.CIRCTCalyxTransforms.dir/GICM.cpp.o tools/circt/lib/Dialect/Calyx/Transforms/CMakeFiles/obj.CIRCTCalyxTransforms.dir/GoInsertion.cpp.o tools/circt/lib/Dialect/Calyx/Transforms/CMakeFiles/obj.CIRCTCalyxTransforms.dir/ClkResetInsertion.cpp.o tools/circt/lib/Dialect/Calyx/Transforms/CMakeFiles/obj.CIRCTCalyxTransforms.dir/RemoveGroups.cpp.o tools/circt/lib/Dialect/Calyx/Transforms/CMakeFiles/obj.CIRCTCalyxTransforms.dir/CalyxHelpers.cpp.o tools/circt/lib/Dialect/Calyx/Transforms/CMakeFiles/obj.CIRCTCalyxTransforms.dir/CalyxLoweringUtils.cpp.o  -Wl,-rpath,/__w/circt/circt/build/lib:  lib/libCIRCTCalyx.so  lib/libCIRCTSupport.so  lib/libCIRCTSV.so  lib/libCIRCTComb.so  lib/libCIRCTHW.so  lib/libMLIRArithmetic.so  lib/libMLIRDialect.so  lib/libMLIRTransforms.so  lib/libMLIRTransformUtils.so  lib/libMLIRRewrite.so  lib/libMLIRPDLToPDLInterp.so  lib/libMLIRPass.so  lib/libMLIRPDLInterp.so  lib/libMLIRPDL.so  lib/libMLIRAnalysis.so  lib/libMLIRInferTypeOpInterface.so  lib/libMLIRCallInterfaces.so  lib/libMLIRControlFlowInterfaces.so  lib/libMLIRDataLayoutInterfaces.so  lib/libMLIRSideEffectInterfaces.so  lib/libMLIRViewLikeInterface.so  lib/libMLIRLoopLikeInterface.so  lib/libMLIRCopyOpInterface.so  lib/libMLIRIR.so  lib/libMLIRSupport.so  -lpthread  lib/libLLVMCore.so  lib/libLLVMSupport.so  -Wl,-rpath-link,/__w/circt/circt/build/lib && :
ld.lld: error: undefined symbol: mlir::detail::TypeIDResolver<mlir::memref::LoadOp, void>::id
>>> referenced by CalyxLoweringUtils.cpp
>>>               tools/circt/lib/Dialect/Calyx/Transforms/CMakeFiles/obj.CIRCTCalyxTransforms.dir/CalyxLoweringUtils.cpp.o:(mlir::Op<mlir::memref::LoadOp, mlir::OpTrait::ZeroRegions, mlir::OpTrait::OneResult, mlir::OpTrait::OneTypedResult<mlir::Type>::Impl, mlir::OpTrait::ZeroSuccessors, mlir::OpTrait::AtLeastNOperands<1u>::Impl, mlir::OpTrait::OpInvariants, mlir::OpTrait::MemRefsNormalizable, mlir::MemoryEffectOpInterface::Trait>::classof(mlir::Operation*))

ld.lld: error: undefined symbol: mlir::detail::TypeIDResolver<mlir::memref::StoreOp, void>::id
>>> referenced by CalyxLoweringUtils.cpp
>>>               tools/circt/lib/Dialect/Calyx/Transforms/CMakeFiles/obj.CIRCTCalyxTransforms.dir/CalyxLoweringUtils.cpp.o:(mlir::Op<mlir::memref::StoreOp, mlir::OpTrait::ZeroRegions, mlir::OpTrait::ZeroResults, mlir::OpTrait::ZeroSuccessors, mlir::OpTrait::AtLeastNOperands<2u>::Impl, mlir::OpTrait::OpInvariants, mlir::OpTrait::MemRefsNormalizable, mlir::MemoryEffectOpInterface::Trait>::classof(mlir::Operation*))
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Any help fixing this linking error would be greatly appreciated.

**Comment by mortbopet** [comment]
> Any help fixing this linking error would be greatly appreciated.

Adding `MLIRMemRef` to `circt/lib/Dialect/Calyx/Transforms/CMakeLists.txt` might fix this.

File: include/circt/Dialect/Calyx/CalyxLoweringUtils.h [hunk 1]
```diff
@@ -88,6 +89,43 @@ class WhileOpInterface {
...
+
+// If the provided type is an index type, converts it to i32, else, returns the
+// unmodified type.
+Type convIndexType(PatternRewriter &rewriter, Type type);
+
+// Creates a new calyx::CombGroupOp or calyx::GroupOp group within compOp.
+template <typename TGroup>
+TGroup createGroup(PatternRewriter &rewriter, calyx::ComponentOp compOp,
+                   Location loc, Twine uniqueName) {
+
```
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxLoweringUtils.h:123]: newline

File: lib/Dialect/Calyx/Transforms/CalyxHelpers.cpp [hunk 2]
```diff
@@ -33,5 +35,46 @@ hw::ConstantOp createConstant(Location loc, OpBuilder &builder,
...
+  ports.push_back({
+      rewriter.getStringAttr("done"),
+      rewriter.getI1Type(),
+      calyx::Direction::Output,
+      getMandatoryPortAttr(ctx, "done"),
+  });
+}
+
+unsigned handleZeroWidth(int64_t dim) {
+  return std::max(llvm::Log2_64_Ceil(dim), 1U);
```
Comment (mortbopet) [inline lib/Dialect/Calyx/Transforms/CalyxHelpers.cpp:76]: remember to update https://github.com/llvm/circt/issues/2909

**Review by mortbopet — APPROVED** [review]

**Comment by cgyurgyik** [comment]
> > Any help fixing this linking error would be greatly appreciated.
> 
> Adding `MLIRMemRef` to `circt/lib/Dialect/Calyx/Transforms/CMakeLists.txt` might fix this.

Thank you ! That fixed things.

**Pull request merged**